recipes-kernel/linux/linux-raspberrypi/0005-Make-ftrace-able-to-mount-data-without-reboot.patch

```diff
@@ -0,0 +1,247 @@
+
+diff --git a/fs/pstore/ftrace.c b/fs/pstore/ftrace.c
+index 5c0450701293..9956c3df6994 100644
+--- a/fs/pstore/ftrace.c
++++ b/fs/pstore/ftrace.c
+@@ -23,6 +23,29 @@
+ /* This doesn't need to be atomic: speed is chosen over correctness here. */
+ static u64 pstore_ftrace_stamp;
+ 
++#define PSTORE_USE_PMU (1)
++#define SEL4_TRACEBUFFER (1)
++
++#define read_sysreg(r) ({                                               \
++	u64 __val;                                                      \
++	asm volatile("mrs %0, " __stringify(r) : "=r" (__val));         \
++	__val;                                                          \
++})
++
++static u64 pstore_ftrace_read_pmccntr(void)
++{
++	return read_sysreg(pmccntr_el0);
++}
++
++static u64 pstore_ftrace_get_timestamp(void)
++{
++#if PSTORE_USE_PMU
++	return pstore_ftrace_read_pmccntr();
++#else
++	return pstore_ftrace_stamp++;
++#endif
++}
++
+ static void notrace pstore_ftrace_call(unsigned long ip,
+ 				       unsigned long parent_ip,
+ 				       struct ftrace_ops *op,
+@@ -44,7 +67,7 @@ static void notrace pstore_ftrace_call(unsigned long ip,
+ 
+ 	rec.ip = ip;
+ 	rec.parent_ip = parent_ip;
+-	pstore_ftrace_write_timestamp(&rec, pstore_ftrace_stamp++);
++	pstore_ftrace_write_timestamp(&rec, pstore_ftrace_get_timestamp());
+ 	pstore_ftrace_encode_cpu(&rec, raw_smp_processor_id());
+ 	psinfo->write(&record);
+ 
+@@ -74,10 +97,16 @@ static ssize_t pstore_ftrace_knob_write(struct file *f, const char __user *buf,
+ 		goto out;
+ 
+ 	if (on) {
++		/* Keep this for debug purposes */
++		pstore_ftrace_call((unsigned long) pstore_ftrace_knob_write,
++				   0x0000, &pstore_ftrace_ops, NULL);
++
+ 		ftrace_ops_set_global_filter(&pstore_ftrace_ops);
+ 		ret = register_ftrace_function(&pstore_ftrace_ops);
+ 	} else {
+ 		ret = unregister_ftrace_function(&pstore_ftrace_ops);
++		/* Reset ftrace buffers and update stored data, #TODO: rename? */
++		ramoops_update_stored_ftrace();
+ 	}
+ 
+ 	if (ret) {
+diff --git a/fs/pstore/internal.h b/fs/pstore/internal.h
+index 7fb219042f13..eb3021332573 100644
+--- a/fs/pstore/internal.h
++++ b/fs/pstore/internal.h
+@@ -46,6 +46,8 @@ extern int	pstore_mkfile(struct dentry *root,
+ extern void	pstore_record_init(struct pstore_record *record,
+ 				   struct pstore_info *psi);
+ 
++void ramoops_update_stored_ftrace(void);
++
+ /* Called during pstore init/exit. */
+ int __init	pstore_init_fs(void);
+ void __exit	pstore_exit_fs(void);
+diff --git a/fs/pstore/ram.c b/fs/pstore/ram.c
+index ca6d8a867285..b67b28f01ce1 100644
+--- a/fs/pstore/ram.c
++++ b/fs/pstore/ram.c
+@@ -878,6 +878,28 @@ static int ramoops_remove(struct platform_device *pdev)
+ 	return 0;
+ }
+ 
++void ramoops_update_stored_ftrace(void)
++{
++    int i;
++    struct ramoops_context *cxt = &oops_cxt;
++
++    pstore_put_backend_records(&cxt->pstore);
++
++    for (i = 0; i != cxt->max_ftrace_cnt; ++i) {
++        struct persistent_ram_zone *prz = cxt->fprzs[i];
++
++	if (!prz) {
```

Comment:
These lines above and below this block seem to have different indentation? Mixed tabs and spaces?

Reply:
Yeap, we actually don't need this part right now, probably it would be better to remove it at all from this PR. And keep somewhere else.